go/runtime/registry/handlers.go

```diff
@@ -0,0 +1,189 @@
+	var rs syncer.ReadSyncer
+	switch rq.Endpoint {
+	case protocol.HostStorageEndpointRuntime:
+		// Runtime storage.
+		rs = h.runtime.Storage()
+	case protocol.HostStorageEndpointConsensus:
+		// Consensus state storage.
+		rs = h.consensus.State()
+	default:
+		return nil, errEndpointNotSupported
+	}
+
+	var rsp *storage.ProofResponse
+	var err error
+	switch {
+	case rq.SyncGet != nil:
+		rsp, err = rs.SyncGet(ctx, rq.SyncGet)
+	case rq.SyncGetPrefixes != nil:
+		rsp, err = rs.SyncGetPrefixes(ctx, rq.SyncGetPrefixes)
+	case rq.SyncIterate != nil:
+		rsp, err = rs.SyncIterate(ctx, rq.SyncIterate)
+	default:
+		return nil, errMethodNotSupported
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &protocol.HostStorageSyncResponse{ProofResponse: rsp}, nil
+}
+
+func handlerHostLocalStorageGetRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostLocalStorageGetRequest) (*protocol.HostLocalStorageGetResponse, error) {
+	value, err := h.runtime.LocalStorage().Get(rq.Key)
+	if err != nil {
+		return nil, err
+	}
+	return &protocol.HostLocalStorageGetResponse{Value: value}, nil
+}
+
+func handlerHostLocalStorageSetRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostLocalStorageSetRequest) (*protocol.Empty, error) {
+	if err := h.runtime.LocalStorage().Set(rq.Key, rq.Value); err != nil {
+		return nil, err
+	}
+	return &protocol.Empty{}, nil
+}
+
+func handlerHostFetchConsensusBlockRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostFetchConsensusBlockRequest) (*protocol.HostFetchConsensusBlockResponse, error) {
+	lb, err := h.consensus.GetLightBlock(ctx, int64(rq.Height))
+	if err != nil {
+		return nil, err
+	}
+	return &protocol.HostFetchConsensusBlockResponse{
+		Block: *lb,
+	}, nil
+}
+
+func handlerHostFetchGenesisHeightRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostFetchGenesisHeightRequest) (*protocol.HostFetchGenesisHeightResponse, error) {
+	doc, err := h.consensus.GetGenesisDocument(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &protocol.HostFetchGenesisHeightResponse{
+		Height: uint64(doc.Height),
+	}, nil
+}
+
+func handlerHostFetchTxBatchRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostFetchTxBatchRequest) (*protocol.HostFetchTxBatchResponse, error) {
+	txPool, err := h.env.GetTxPool(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	batch := txPool.GetSchedulingExtra(rq.Offset, rq.Limit)
+	raw := make([][]byte, 0, len(batch))
+	for _, tx := range batch {
+		raw = append(raw, tx.Raw())
+	}
+
+	return &protocol.HostFetchTxBatchResponse{
+		Batch: raw,
+	}, nil
+}
+
+func handlerHostProveFreshnessRequest(ctx context.Context, h *runtimeHostHandler, rq *protocol.HostProveFreshnessRequest) (*protocol.HostProveFreshnessResponse, error) {
+	// Subscribe before submit as we don't want to miss the block with the transaction.
```

Comment:
This could also be part of the consensus backend interface, e.g. something like `SubmitTxWithProof`. Especially given that the `SubmitTx` method already subscribes to and waits for results:
https://github.com/oasisprotocol/oasis-core/blob/b64e8b5a08c27a31066b47e232b09087b8c51f70/go/consensus/tendermint/full/full.go#L247

This would also make it easier to abstract out the proof type (e.g. use `[]byte` as mentioned earlier).